### PR TITLE
migrate circle 1.0 to 2.0

### PR DESCRIPTION
**Pull request by yongpingchen**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            tags:
+              ignore: /.*/
+  build_deploy:
+    jobs:
+      - build:
+          filters:
+            branches:
+              ignore: /.*/
+            tags:
+              only: /.*/
```
Comment (satoshikumano) [inline .circleci/config.yml]: Why not apply pattern same as `deployment.filters.tags` ?

**Review by satoshikumano — COMMENTED** [review]
Left comment

**Comment by yongpingchen** [comment]
Changes since mob programming:
- need to install doxygen and cmake in docker image of each job
- cache submodules in build job, and test job uses the saved cache
- add deployment job, and apply it to valid tag
- separate workflows: 
  - build_test: apply for all branch, when there is commit 
  - build_deploy: apply to valid tags, when deployment.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,74 @@
+version: 2
+jobs:
+  build:
+    docker:
+      - image: buildpack-deps:trusty
+    steps:
+      - checkout
+      - run:
+          name: install dependencies
+          command: sudo apt update && sudo apt install doxygen && echo y | sudo apt install cmake3
```
Comment (naoki-akimoto) [inline .circleci/config.yml]: commandは多分ここと同じようにすれば"&&"でつながず分割して書けるのではないかと思います。
https://circleci.com/docs/2.0/migrating-from-1-2/#interpolating-environment-variables

File: .circleci/config.yml (hunk 1 above)
Comment (yongpingchen) [reply .circleci/config.yml]: fixed on commit: 4f7a723abfab4c8d78281514b20cb5944d5ed716

File: .circleci/config.yml (hunk 2 above)
Comment (yongpingchen) [reply .circleci/config.yml]: fixed with commit: 10723c8

**Review by satoshikumano — APPROVED** [review]
LGTM!

**Review by naoki-akimoto — APPROVED** [review]
LGTM!

**Review by nfukuzaki — APPROVED** [review]
LGTM

**Pull request merged**